packages/solarwinds-apm/src/storage.ts

```diff
@@ -0,0 +1,163 @@
+limitations under the License.
+*/
+
+import {
+  type Context,
+  createContextKey,
+  diag,
+  type Span,
+} from "@opentelemetry/api"
+import * as sdk from "@opentelemetry/sdk-trace-base"
+
+/**
+ * Creates a global value shared between ESM and CommonJS contexts
+ *
+ * @param id - Descriptive unique ID for the global
+ * @param init - Initializer for the global if it needs to be
+ *
+ * @returns Shared global
+ */
+export function global<const T>(id: string, init: () => T): T {
+  const key = Symbol.for(`solarwinds-apm global storage / ${id}`)
+  let storage = Reflect.get(globalThis, key) as T | undefined
+
+  if (!storage) {
+    storage = init()
+    Reflect.defineProperty(globalThis, key, {
+      value: storage,
+      configurable: false,
+      enumerable: false,
+      writable: false,
+    })
+  }
+
+  return storage
+}
+
+/** Typed wrapper around the OTel {@link Context} API for a specific key */
+export interface ContextStorage<T> {
+  /**
+   * Gets the value for the key
+   *
+   * Note that while the context object itself is immutable,
+   * reference types stored in it are not. This means it is possible
+   * to modify the return value of this function in-place if it is
+   * a reference type.
+   */
+  get(context: Context): T | undefined
+  /**
+   * Sets the value for the key
+   *
+   * Note that the context object is immutable and this function
+   * returns an updated context without modifying the given one in-place.
+   * in-place
+   */
+  set(context: Context, value: T): Context
+  /**
+   * Deletes the value for the key
+   *
+   * Note that the context object is immutable and this function
+   * returns an updated context without modifying the given one in-place.
+   */
+  delete(context: Context): Context
+}
+
+/**
+ * Typed OTel {@link Span} attached storage API for a specific key
+ *
+ * This makes it possible to attach arbitrary data to OTel spans
+ * without altering the span object.
+ *
+ * The stored value will be accessible throughout the entire
+ * lifecycle of the span from the moment it is set until the span
+ * finishes being exported.
+ *
+ * Unlike the previous implementation which used the span ID as a key
+ * and needed to be carefully cleared at the right times to avoid leaking
+ * memory, this uses a WeakMap with the span object itself as a key
+ * which ensures the stored values are kept at least as long as they need
+ * to be accessed.
+ *
+ * However it it still useful to manually clear unused entries to avoid
+ * using up memory unnecessarily while waiting for the next GC cycle.
+ */
+export interface SpanStorage<T> {
+  get(span: Span | sdk.Span | sdk.ReadableSpan): T | undefined
+  set(span: Span | sdk.Span | sdk.ReadableSpan, value: T): boolean
+  delete(span: Span | sdk.Span | sdk.ReadableSpan): boolean
+}
+
+const GLOBAL_SPAN_STORAGE = global(
+  "solarwinds-apm span storage",
+  () => new WeakMap<sdk.Span, Map<symbol, unknown>>(),
```

Comment:
Starting from the beginning for me! 😄 I think the WeakMap (Python equivalent is `weakref`?) is a nice idea for this application.

Does this use full Span objects as keys?

What are the values under each key? (the `Map<symbol, unknown>` part)

Reply:
Yes, there is a global WeakMap (I believe the Python equivalent would be `weakref.WeakKeyMap` ?) that maps full span objects (technically their pointer) to a second map that maps keys to arbitrary values. When calling `spanStorage` a key for the inner map is created and linked to the returned `SpanStorage` instance. So on access the Span is used to index the outer WeakMap, which retrieves the inner map that is then indexed with the linked key. This is to make it possible to have multiple independent instances of `SpanStorage` that don't conflict with each other without creating a bunch of global WeakMaps.

Reply:
`span` is the key for the outer map, the ID created by the call to `spanStorage` (a unique symbol) is the key for the inner map, and `parentSpan` is the value set in the inner map

Reply:
Ohhhhh JS `Symbol` constructor is unique! Got it, thank you.